zkvm/src/predicate.rs

```diff
@@ -181,75 +151,195 @@ impl Into<CompressedRistretto> for Predicate {
     }
 }
 
-#[cfg(test)]
-mod tests {
-    use super::*;
-    use bulletproofs::PedersenGens;
+impl PredicateTree {
+    /// Creates new predicate tree with a verification key and a list of programs
+    pub fn new(
+        key: Option<VerificationKey>,
+        progs: Vec<Program>,
+        blinding_key: [u8; 32],
+    ) -> Result<Self, VMError> {
+        let key = key.unwrap_or_else(|| Predicate::unsignable_key());
+        let leaves = Self::commit_leaves(&progs, blinding_key);
+        let root = MerkleTree::root(b"ZkVM.taproot", &leaves);
+
+        // P = X + h(X, M)*G
+        let adjustment_factor = Predicate::commit_taproot(key.0.as_bytes(), &root);
+        let precomputed_point = {
+            let h = adjustment_factor;
+            let x = key.0.decompress().ok_or(VMError::InvalidPoint)?;
+            let p = x + h * PedersenGens::default().B;
+            p.compress()
+        };
 
-    fn bytecode(prog: &Program) -> Vec<u8> {
-        let mut prog_vec = Vec::new();
-        prog.encode(&mut prog_vec);
-        prog_vec
+        Ok(Self {
+            leaves,
+            blinding_key,
+            key,
+            precomputed_point,
+            adjustment_factor,
+        })
     }
 
-    #[test]
-    fn valid_program_commitment() {
-        let prog = Program::build(|p| p.drop());
-        let pred = Predicate::unblinded_program(prog.clone());
-        let op = pred.prove_program_predicate(&bytecode(&prog), &[]);
-        assert!(op.verify().is_ok());
+    /// Returns the adjustment factor for signing
+    pub fn adjustment_factor(&self) -> Scalar {
```

Comment:
probably add a comment that instead of this API we'd rather return a "key witness" object like musig::Multikey and  the adjustment factor will be stored right there.